### PR TITLE
Refactored Data to be more robust.  Updated UI to be compatible.

### PR DESCRIPTION
**Pull request by JohnLeeroy**

### CODE REVIEW

File: app/src/main/java/com/codefororlando/petadoption/data/util/AnimalJsonDeserializer.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+public class AnimalJsonDeserializer implements JsonDeserializer<Animal> {
```
Comment (ToxicBakery) [inline app/src/main/java/com/codefororlando/petadoption/data/util/AnimalJsonDeserializer.java]: Why bother with the custom deserializer?

File: app/src/main/java/com/codefororlando/petadoption/data/util/RetrievableImplJsonDeserializer.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import com.codefororlando.petadoption.data.impl.RetrievableImpl;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParseException;
+
+import java.lang.reflect.Type;
+
+public class RetrievableImplJsonDeserializer implements JsonDeserializer<RetrievableImpl> {
```
Comment (ToxicBakery) [inline app/src/main/java/com/codefororlando/petadoption/data/util/RetrievableImplJsonDeserializer.java]: Same as before, just asking for bugs

**Comment by ToxicBakery** [comment]
So the more appropriate explanation against the deserializers is that they are specific your custom JSON model which will almost undoubtedly end up not matching the API. Even if they do end up matching, the custom deserializers are really only necessary when normal deserialization fails, usually for a complex reason, or because you have a *lot* of JSON to serialize or deserialize. Even then it would likely be best to move away from JSON to something faster vs writing deserializers as they are just extra work.

**Comment by JohnLeeroy** [comment]
@ToxicBakery 
Thanks for the review, I removed the deserializers and modified the json file to deserialize nicely.  I just had to add a custom deserializer for Uri.  

I agree that custom serializer/deserializer is undesirable in most cases.  I dislike the idea of maintaining them.  It seems I was lazy with how I structured my json.

**Comment by ToxicBakery** [comment]
Yeah, Uri is a good case for when you have to use one.

**Pull request merged**